ledger/participant-integration-api/src/main/scala/platform/indexer/parallel/ParallelIndexerFactory.scala

```diff
@@ -115,16 +98,160 @@ object ParallelIndexerFactory {
+  )
+
+  def inputMapper(
+      metrics: Metrics,
+      toDbDto: Offset => Update => Iterator[DBDTOV1],
+  ): Iterable[((Offset, Update), Long)] => RunningBatch[Vector[DBDTOV1]] = { input =>
+    metrics.daml.parallelIndexer.inputMapping.batchSize.update(input.size)
+    val batch = input.iterator.flatMap { case ((offset, update), _) =>
+      toDbDto(offset)(update)
+    }.toVector
+    RunningBatch(
+      lastOffset = input.last._1._1,
+      lastSeqEventId = 0, // will be filled later in the sequential step
+      lastConfig = batch.reverseIterator.collectFirst {
+        case c: DBDTOV1.ConfigurationEntry if c.typ == JdbcLedgerDao.acceptType => c.configuration
+      },
+      batch = batch,
+      batchSize = input.size,
+      averageStartTime = input.view.map(_._2 / input.size).sum,
+    )
+  }
+
+  def seqMapperZero(initialSeqId: Long): RunningBatch[Vector[DBDTOV1]] =
+    RunningBatch(
+      lastOffset = null,
+      lastSeqEventId = initialSeqId, // this is the only property of interest in the zero element
+      lastConfig = None,
+      batch = Vector.empty,
+      batchSize = 0,
+      averageStartTime = 0,
+    )
+
+  def seqMapper(metrics: Metrics)(
+      previous: RunningBatch[Vector[DBDTOV1]],
+      current: RunningBatch[Vector[DBDTOV1]],
+  ): RunningBatch[Vector[DBDTOV1]] = {
+    var eventSeqId = previous.lastSeqEventId
+    val batchWithSeqIds = current.batch.map {
+      case dbDto: DBDTOV1.EventCreate =>
+        eventSeqId += 1
+        dbDto.copy(event_sequential_id = eventSeqId)
+
+      case dbDto: DBDTOV1.EventExercise =>
+        eventSeqId += 1
+        dbDto.copy(event_sequential_id = eventSeqId)
+
+      case dbDto: DBDTOV1.EventDivulgence =>
+        eventSeqId += 1
+        dbDto.copy(event_sequential_id = eventSeqId)
+
+      case notEvent => notEvent
+    }
+    val nowNanos = System.nanoTime()
+    metrics.daml.parallelIndexer.inputMapping.duration.update(
+      (nowNanos - current.averageStartTime) / current.batchSize,
+      TimeUnit.NANOSECONDS,
+    )
+    current.copy(
+      lastSeqEventId = eventSeqId,
+      batch = batchWithSeqIds,
+      averageStartTime = nowNanos, // setting start time to the start of the next stage
+    )
+  }
+
+  def batcher[DB_BATCH](
+      batchF: Vector[DBDTOV1] => DB_BATCH,
+      metrics: Metrics,
+  ): RunningBatch[Vector[DBDTOV1]] => RunningBatch[DB_BATCH] = { inBatch =>
+    val dbBatch = batchF(inBatch.batch)
+    val nowNanos = System.nanoTime()
+    metrics.daml.parallelIndexer.batching.duration.update(
+      (nowNanos - inBatch.averageStartTime) / inBatch.batchSize,
+      TimeUnit.NANOSECONDS,
+    )
+    inBatch.copy(
+      batch = dbBatch,
+      averageStartTime = nowNanos,
+    )
+  }
+
+  def ingester[DB_BATCH](
+      ingestFunction: (Connection, DB_BATCH) => Unit,
+      dbDispatcher: DbDispatcher,
+      metrics: Metrics,
+  )(implicit
+      loggingContext: LoggingContext // TODO is that right so? shouldn't we prep something here?
```

Comment:
What would you like to prep? Add some info to the logging context, like the batch size?

Reply:
I have no clue :) essentially this is the same what I was explaining above.

Reply:
I would only add values to the logging context if we think they might be useful for troubleshooting issues later. `batch.lastSeqEventId` is the closest to a unique ID of the current batch, which might be helpful for correlating log lines. You could consider adding the following line, but I don't think it's very important:

```
withEnrichedLoggingContext("lastSeqEventId" -> batch.lastSeqEventId) { implicit loggingContext =>
```

Reply:
What I had in mind, whether we already had something enriched already, which we were relying on...
The lastqlEventId can be actually the same for many batches, this is only for events. Maybe use the hex-offset?
I suggest turning this around: let's have needs from logging and then implement it here.
On log correlation: that is usually quite hard, due to batching. For example to track how a submission is travelling trough a system: for that we would need to do is to be able to correlate batches to all submissions which they relate to :thinking: